tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -548,46 +593,60 @@ struct BatchNarrowMatrixTransposeDispatcher {
                    const Dimension<3>& input_dims, T* output) {
     static_assert(
         (TileLongSide & (TileLongSide - 1)) == 0,
         "The length of the longer side of the tile is always a power of 2.");
+    bool request_satisfied = max(tile_size_i, tile_size_j) <= TileLongSide &&
+                             min(tile_size_i, tile_size_j) <= TileShortSide;
 
-    constexpr int NumThreads = TileLongSide;
-    if (tile_size_i <= TileLongSide && tile_size_j <= TileShortSide) {
-      SwapDimension1And2InTensor3UsingTiles<T, NumThreads, TileLongSide,
-                                            TileShortSide>
-          <<<total_tiles_count, NumThreads, 0, d.stream()>>>(input, input_dims,
-                                                             output);
-    } else {
-      SwapDimension1And2InTensor3UsingTiles<T, NumThreads, TileShortSide,
-                                            TileLongSide>
-          <<<total_tiles_count, NumThreads, 0, d.stream()>>>(input, input_dims,
-                                                             output);
+    if (request_satisfied) {
+      LaunchBatchNarrowMatrixTransposeKernel<T, TileLongSide, TileShortSide>(
+          d, tile_size_i, tile_size_j, total_tiles_count, input, input_dims,
+          output);
     }
+
+    // If the execution reaches here, then the kernel was not launched; since
+    // we are on the boundary of the performant subspace but not on the
+    // frontier, we increment the short dimension and try again.
+    BatchNarrowMatrixTransposeDispatcher<
+        T, TileLongSide, TileShortSide + 1>::DoIt(d, tile_size_i, tile_size_j,
+                                                  total_tiles_count, input,
+                                                  input_dims, output);
   }
 };
 
 template <typename T, int TileLongSide, int TileShortSide>
 struct BatchNarrowMatrixTransposeDispatcher<
     T, TileLongSide, TileShortSide,
-    typename std::enable_if<TileSizePastFrontier(TileLongSide, TileShortSide,
-                                                 sizeof(T)),
+    typename std::enable_if<TileSizeOnFrontier(TileLongSide, TileShortSide,
+                                               sizeof(T)),
                             void>::type> {
   static void DoIt(const GPUDevice& d, int tile_size_i, int tile_size_j,
                    int total_tiles_count, const T* input,
                    const Dimension<3>& input_dims, T* output) {
-    assert(false &&
-           "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "
-           "launch configuration. ");
+    static_assert(
+        (TileLongSide & (TileLongSide - 1)) == 0,
+        "The length of the longer side of the tile is always a power of 2.");
+
+    LaunchBatchNarrowMatrixTransposeKernel<T, TileLongSide, TileShortSide>(
+        d, tile_size_i, tile_size_j, total_tiles_count, input, input_dims,
+        output);
   }
 };
 
+template <typename T, int TileLongSide, int TileShortSide>
```

Comment:
Maybe add

  // Base case -- no `DoIt` function because it's illegal to do transposes when the tile sizes are outside of the boundary.

Although, just to check: Do you need this specialization?  I'd guess you maybe don't, because it should never be instantiated anymore.

Reply:
Indeed no need!